### PR TITLE
refactor(orchestrator): score native episodes

### PR DESCRIPTION
**Pull request by hallerite**
## Summary

Companion: PrimeIntellect-ai/verifiers#2409, merged and published in `verifiers==0.3.1.dev59`; the submodule is pinned to release commit `4bcb48e55`.

Makes credit assignment and admission operate on native `vf.Episode` objects. The verifier graph is the scored artifact; `TrainingSample` is now only the final trainer transport compiled after admission.

- Algorithms receive episodes, annotating verifier nodes with advantages, reference logprobs, and named loss weights.
- Task samplers observe every finalized episode group; admission gates inspect the same scored episodes and compose with AND.
- The train sink scores the episode, scores the group, evaluates admission, and only then compiles admitted traces into trainer samples. `PreparedTrace` and `PreparedGroup` are no longer needed.
- The dispatcher stamps the full task, environment, structured `GroupInfo`, actual dispatch step, train/eval work kind, and policy span directly on the episode.
- Returned task key and content hash are validated against the dispatched task; class names are not compared because the request protocol transports task data, not its Python class.
- Unexpected dispatcher or watcher termination is propagated to the orchestrator instead of leaving the main loop waiting on an output queue that can no longer advance.
- Prime requires environment, group, and work provenance instead of repairing or defaulting malformed episodes.
- Curriculum checkpoints must exactly match the sampler, gates, and configured environments; incompatible state fails immediately.
- Exact sampled trace ids drive batching and metrics, including partial groups and post-batch zero-advantage filtering.
- Rollout JSONL is exactly one native episode per line, preserving multi-trace and trace-less failure episodes.
- Difficulty-pool sampling continues to use the sampler's `observe()` hook, while `AdvRangeGate` reads the native advantages written by the algorithm.

## Public hooks

```python
class Algorithm:
    async def score_episode(self, episode: vf.Episode) -> None: ...
    async def score_group(self, episodes: list[vf.Episode]) -> None: ...

class TaskSampler(Iterator[vf.Task]):
    def __next__(self) -> vf.Task: ...
    def observe(self, group: list[vf.Episode]) -> None: ...
    def state_dict(self) -> dict[str, Any]: ...
    def load_state_dict(self, state_dict: dict[str, Any]) -> None: ...

class AdmissionGate:
    def admit(self, group: list[vf.Episode]) -> bool: ...
    def state_dict(self) -> dict[str, Any]: ...
    def load_state_dict(self, state_dict: dict[str, Any]) -> None: ...
```

## Validation

- `pytest -q tests/unit/orchestrator --ignore=tests/unit/orchestrator/test_qwen3_vl_e2e.py` — 79 passed
- `uv lock --check`; editable verifiers import reports `0.3.1.dev59`
- verifier v1 suite against the pinned submodule — 70 passed
- `ruff check` / `ruff format --check` on the changed surface
- `ty check` on the changed orchestrator surface
- in-memory train-sink proof confirmed gates see native advantages before two admitted samples are compiled
- earlier curriculum revision: two-GPU reverse-text RL smoke completed 2/2 optimizer steps, and checkpoint-resume restored sampler RNG and gate state

---

> [!NOTE]
> **High Risk**
> This is a core orchestrator rewrite of credit assignment, admission, dispatch provenance, metrics, and checkpoints. Bugs here can silently change training signal, drop/keep the wrong traces, or break resume.
> 
> **Overview**
> Makes credit assignment and admission run on native `vf.Episode` graphs. `TrainingSample` is compiled only after a group is admitted.
> 
> **Algorithms** now override `score_episode` / `score_group` and write advantages, reference logprobs, and named loss weights onto verifier nodes (`assign_advantages`, `assign_reference_logprobs`). The train sink scores the episode, scores the group, runs curriculum gates, then flattens admitted traces.
> 
> **Dispatch and logging** emit one episode (not a list of `Rollout`s). Provenance (env, group, train/eval work, policy span, dispatch step) is stamped on the episode; JSONL is one native episode per line, including trace-less failures. Metrics and batching key off exact sampled trace ids.
> 
> Curriculum checkpoints must match configured samplers, gates, and envs exactly. Unexpected dispatcher/watcher exit is raised instead of hanging the main loop.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9332485bf072bec6650bb9bfe0603140d918999e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/algorithms.md [hunk 1]
```diff
@@ -158,20 +158,20 @@ At runtime, each env's resolved config builds two objects: a `Sampler` (`prime_r
 | `algo.type` | Class | hook(s) — stage |
 |---|---|---|
 | `grpo` | `GRPOAlgorithm` | `score_group`: group-norm credit (optional length penalty) |
-| `echo` | `EchoAlgorithm` | `score_rollout`: weighted ce on observation tokens; `score_group`: group-norm credit (inherited) |
+| `echo` | `EchoAlgorithm` | `score_trace`: weighted ce on observation tokens; `score_group`: group-norm credit (inherited) |
```
Comment (mikasenghaas) [inline docs/algorithms.md]: what about scoring a episode?

File: docs/algorithms.md [hunk 2]
```diff
@@ -158,20 +158,20 @@ At runtime, each env's resolved config builds two objects: a `Sampler` (`prime_r
...
-| `opd` | `OPDAlgorithm` | `score_rollout`: own-context prefill under the teacher |
-| `opsd` | `OPSDAlgorithm` | `score_rollout`: demo-conditioned prefill under the live policy |
+| `opd` | `OPDAlgorithm` | `score_trace`: own-context prefill under the teacher |
+| `opsd` | `OPSDAlgorithm` | `score_trace`: demo-conditioned prefill under the live policy |
 | `sft` | `SFTDistillAlgorithm` | no credit assignment; CE on sampled tokens |
 
-Each class owns its hooks outright — reading one top to bottom reads the algorithm, and everything on the class is an override point. The two hooks are one scope-and-timing ladder — the wider scope is unlocked by a later barrier, so the two axes coincide. Each is handed the `Rollout` directly — the env's typed trace (`reward`, `nodes`, `num_turns`, ...) with `samples` attached, plus `assign_advantages` to write credit:
+Each class owns its hooks outright. Verifier episodes carry rollout provenance; trainer preparation stays separate as `PreparedTrace = list[TrainingSample]`, keyed by trace id in a `PreparedGroup`.
```
Comment (mikasenghaas) [inline docs/algorithms.md]: i dont like the `Prepared` naming. maybe just `Train` or `Trainable`?

File: docs/algorithms.md [hunk 3]
```diff
@@ -420,30 +420,33 @@ Both of `kuhn-poker`'s agents late-bind to the run's own model — shared-policy
...
+from prime_rl.orchestrator.algo.routing import assign_advantages
 
 
 class MyAlgorithm(Algorithm):
-    async def score_group(self, group):
-        rewards = torch.tensor([rollout.reward for rollout in group], dtype=torch.float32)
-        advantages = ...  # one value per rollout
-        for rollout, advantage in zip(group, advantages.tolist(), strict=True):
-            rollout.assign_advantages(advantage)
+    async def score_group(self, episodes, prepared):
```
Comment (mikasenghaas) [inline docs/algorithms.md]: what is prepared here? how does episodes differ from prepared? the signature is not very intuitive to curently

File: skills/install/SKILL.md [hunk 4]
```diff
@@ -26,7 +26,7 @@ uv sync                                    # core only
 uv sync --group dev                        # + pytest, ruff, pre-commit
 uv sync --all-extras                       # + extras (flash-attn, flash-attn-cute, …)
 uv sync --all-extras --all-packages        # + all env packages (needed to train on them)
-uv sync --package prime-rl --package gsm8k  # core + just one env
+uv sync --all-extras --package prime-rl --package gsm8k  # accelerator extras + one env
```
Comment (mikasenghaas) [inline skills/install/SKILL.md]: i liked the old comment more (the # core + just one env)

File: skills/training/monitor-run/SKILL.md [hunk 5]
```diff
@@ -140,31 +140,29 @@ curl -s http://localhost:8100/metrics | grep -E "num_requests|gpu_cache_usage"
 # vllm:num_requests_running, vllm:num_requests_waiting, vllm:gpu_cache_usage_perc (→1.0 = KV cache saturated)
 ```
 
-### Traces
+### Episodes
 
 ```
-{run_dir}/rollouts/step_N/{train,eval}/all/traces.jsonl        # appended per rollout as it completes
+{run_dir}/rollouts/step_N/{train,eval}/all/traces.jsonl        # appended per episode as it completes
```
Comment (mikasenghaas) [inline skills/training/monitor-run/SKILL.md:146]: didnt check code but can we double check that this is the exact smae output format as we write during `uv run eval` in vf (should be one episode per line)

File: src/prime_rl/evals/evals.py [hunk 6]
```diff
@@ -336,19 +338,10 @@ async def consume_epoch(self, fired: list[str]) -> None:
         # An env with no examples emits no episodes, so its epoch can never finalize.
         pending = {env_name for env_name in fired if self.eval_sink.batch_size_for(env_name) > 0}
         while pending:
-            episode: list[Rollout] = await self.dispatcher.out_q.get()
-            step = episode[0].eval_step
+            episode = await self.dispatcher.out_q.get()
+            step = episode.run.step if isinstance(episode.run, vf.EvalRunInfo) else None
```
Comment (mikasenghaas) [inline src/prime_rl/evals/evals.py]: i think both should have a step (the step they were dispatched)

File: src/prime_rl/orchestrator/algo/base.py [hunk 7]
```diff
@@ -134,28 +143,33 @@ async def connect(self, reference: FrozenModelConfig) -> InferencePool:
         self.connected_pools.append(pool)
         return pool
 
-    async def score_rollout(self, rollout: Rollout) -> None:
-        """Arrival phase, one rollout, before its group is complete: write
-        rollout-local credit (``rollout.assign_advantages``), observation ce
-        weights (echo), or per-token results from a model — an inference pool
-        connected in :meth:`setup`, or the live policy (opsd). No siblings, no
-        group stats."""
+    async def score_trace(
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/algo/base.py]: wait i get the trace + episode? how does that work?

File: src/prime_rl/orchestrator/types.py [hunk 8]
```diff
@@ -35,17 +33,17 @@ class Progress:
...
 
 
 @dataclass
 class InflightEpisode:
-    """Per-task scheduling state in the dispatcher; one entry per in-flight
-    ``run`` task."""
+    """Scheduling state for one in-flight environment run."""
 
-    kind: RolloutKind
+    kind: RunKind
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/types.py]: i wouldnt call this "run", its all one train run, just two types of work unit. kind is acc fine imo tho a bit generic

File: src/prime_rl/orchestrator/types.py [hunk 9]
```diff
@@ -59,105 +57,42 @@ class GroupState:
...
-
-    @property
-    def is_trainable(self) -> bool:
-        """Whether the rollout carries a training signal — a nonzero advantage on some token. A
-        uniform-reward GRPO group (all-zero advantages) or an unscored rollout has no gradient."""
-        return bool(self.advantages) and any(a != 0.0 for a in self.advantages)
+type PreparedTrace = list[TrainingSample]
+"""Trainer-bound samples produced from one verifier trace."""
+
+type PreparedGroup = dict[str, PreparedTrace]
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/types.py]: not sure i ge this type? if its the group in the grpo sense why do i need to key?

Comment (mikasenghaas) [reply src/prime_rl/orchestrator/types.py]: or is this always len(1)? in this case i feel like tuple would be more natural. when i see dict i think its a collection

File: skills/training/monitor-run/SKILL.md (hunk 5 above)
Comment (hallerite) [reply skills/training/monitor-run/SKILL.md:146]: pretty sure it is. originally my clanker renamed to `episodes.jsonl`, then I asked to keep the name like in vf and hence it changed the comment I guess

File: tests/unit/orchestrator/test_advantage.py [hunk 10]
```diff
@@ -10,8 +11,15 @@
 )
 from prime_rl.orchestrator.algo.grpo import GRPOAlgorithm
 from prime_rl.orchestrator.algo.max_rl import MaxRLAlgorithm
+from prime_rl.orchestrator.algo.routing import assign_advantages
 from prime_rl.orchestrator.trajectories import trace_to_samples
-from prime_rl.orchestrator.types import Rollout
+from prime_rl.transports.rollouts import TrainingSample
+
+
+class PreparedCase(NamedTuple):
```
Comment (mikasenghaas) [inline tests/unit/orchestrator/test_advantage.py]: what is this?

File: src/prime_rl/orchestrator/provenance.py
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/provenance.py:1]: dont think we need a whole file for this

File: tests/unit/orchestrator/test_advantage.py (hunk 10 above)
Comment (hallerite) [reply tests/unit/orchestrator/test_advantage.py]: Codex called it "transition slop"

**Review by mikasenghaas — DISMISSED** [review]

**Review by mikasenghaas — APPROVED** [review]
